### PR TITLE
quagga service: init

### PR DESCRIPTION
**Pull request by tavyc**
###### Motivation for this change

This adds a NixOS module to configure the Quagga IP routing software suite.

Included is a test case which sets up a network of 4 machines and verifies the correct operation of a minimal OSPF configuration; it takes about a minute to run.

Tested on master and in production on multiple machines running release-16.03.
###### Things done
- [x] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/quagga.nix [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+  ###### implementation
+
+  config = mkIf (any isEnabled allServices) {
+
+    environment.systemPackages = [
+      pkgs.quagga               # for the vtysh tool
+    ];
+
+    users.extraUsers = singleton {
```
Comment (joachifm) [inline nixos/modules/services/networking/quagga.nix]: Slightly denser as `users.users.quagga = { ... }`

File: nixos/modules/misc/ids.nix [hunk 2]
```diff
@@ -275,6 +275,8 @@
       gocd-server = 252;
       terraria = 253;
       mattermost = 254;
+      quagga = 255;
```
Comment (joachifm) [inline nixos/modules/misc/ids.nix]: I'm unfamiliar with this service, what is the reason for assigning it a static id?

File: nixos/modules/services/networking/quagga.nix (hunk 1 above)
Comment (tavyc) [inline nixos/modules/services/networking/quagga.nix]: Changed as suggested and rebased, thanks.

File: nixos/modules/misc/ids.nix (hunk 2 above)
Comment (tavyc) [inline nixos/modules/misc/ids.nix]: The Quagga daemons (zebra, ripd, ospfd, ...) run as user 'quagga' / group 'quagga' and extra group 'quaggavty'.
I've assigned static user/group IDs like all the other modules use for daemons.

Comment (joachifm) [inline nixos/modules/misc/ids.nix]: Not all daemons use static ids. Unless there's a specific reason to use static ids, please switch to dynamic ids.

Comment (tavyc) [inline nixos/modules/misc/ids.nix]: There is no specific reason to use static ids, I just assumed it to be good practice for daemons, according to the comment at the top of misc/ids.nix.

I've updated the module to use dynamic ids, and rebased the PR.

Comment (joachifm) [inline nixos/modules/misc/ids.nix]: I see. Using static ids ensures that data is not suddenly owned by a different user/group due to variability in ids across NixOS systems/activations. Whether this matters depends on the nature of the data (is it expected to be portable, are there privacy/security concerns, &c) and the amount of manual intervention involved to resolve inconsistencies.

Another reason to use static ids is if you need to refer to the user/group before the id/name mapping is established.

If you think it is appropriate for this service, by all means use a static id, but keep in mind that it is a finite resource.

Comment (tavyc) [inline nixos/modules/misc/ids.nix]: Thank you for clarifying the ID allocation policy, it makes sense to try to preserve IDs for services with stateful data.

Quagga does not store any stateful data, it uses the 'quagga' user&group just to drop privileges after startup and to control access to its unix sockets in /run/quagga; no other files should have owner 'quagga'. 
So in this specific case the dynamic user IDs are appropriate.

**Comment by joachifm** [comment]
Ran the test locally, everything okay. Thank you.

**Pull request merged**